src/thread_event.c

```diff
@@ -0,0 +1,255 @@
+			wait = event_wait;				\
+		}							\
+	}
+
+	ITERATE_OVER_ALL_EVENTS
+#undef E
+
+	assert(wait <= THREAD_EVENT_MAX_START_WAIT);
+	return wait;
+}
+
+void
+thread_event_trigger(tsd_t *tsd, bool delay_event) {
+	/* usize has already been added to thread_allocated. */
+	uint64_t thread_allocated_after = thread_allocated_get(tsd);
+
+	/* The subtraction is intentionally susceptible to underflow. */
+	uint64_t accumbytes = thread_allocated_after -
+	    thread_allocated_last_event_get(tsd);
+
+	/* Make sure that accumbytes cannot overflow uint64_t. */
+	cassert(THREAD_EVENT_MAX_INTERVAL <=
+	    UINT64_MAX - SC_LARGE_MAXCLASS + 1);
+
+	thread_allocated_last_event_set(tsd, thread_allocated_after);
+	bool allow_event_trigger = !delay_event && tsd_nominal(tsd) &&
+	    tsd_reentrancy_level_get(tsd) == 0;
+	uint64_t wait = thread_event_trigger_batch_update(tsd, accumbytes,
+	    allow_event_trigger);
+	thread_event_adjust_thresholds_helper(tsd, wait);
+
+	thread_event_assert_invariants(tsd);
+
+#define E(event, condition)						\
+	if (condition && event##_event_wait_get(tsd) == 0U) {		\
+		assert(allow_event_trigger);				\
+		thread_##event##_event_handler(tsd);			\
+	}
+
+	ITERATE_OVER_ALL_EVENTS
+#undef E
+
+	thread_event_assert_invariants(tsd);
+}
+
+void
+thread_event_rollback(tsd_t *tsd, size_t diff) {
+	thread_event_assert_invariants(tsd);
+
+	if (diff == 0U) {
+		return;
+	}
+
+	uint64_t thread_allocated = thread_allocated_get(tsd);
+	/* The subtraction is intentionally susceptible to underflow. */
+	uint64_t thread_allocated_rollback = thread_allocated - diff;
+	thread_allocated_set(tsd, thread_allocated_rollback);
+
+	uint64_t last_event = thread_allocated_last_event_get(tsd);
+	/* Both subtractions are intentionally susceptible to underflow. */
+	if (thread_allocated_rollback - last_event <=
+	    thread_allocated - last_event) {
+		thread_event_assert_invariants(tsd);
+		return;
+	}
+
+	thread_allocated_last_event_set(tsd, thread_allocated_rollback);
+
+	/* The subtraction is intentionally susceptible to underflow. */
+	uint64_t wait_diff = last_event - thread_allocated_rollback;
+	assert(wait_diff <= diff);
+
+#define E(event, condition)						\
+	if (condition) {						\
+		uint64_t event_wait = event##_event_wait_get(tsd);	\
+		assert(event_wait <= THREAD_EVENT_MAX_START_WAIT);	\
+		if (event_wait > 0U) {					\
+			if (wait_diff >					\
+			    THREAD_EVENT_MAX_START_WAIT - event_wait) {	\
+				event_wait =				\
+				    THREAD_EVENT_MAX_START_WAIT;	\
+			} else {					\
+				event_wait += wait_diff;		\
+			}						\
+			assert(event_wait <=				\
+			    THREAD_EVENT_MAX_START_WAIT);		\
+			event##_event_wait_set(tsd, event_wait);	\
+		}							\
+	}
+
+	ITERATE_OVER_ALL_EVENTS
+#undef E
+
+	thread_event_update(tsd);
+}
+
+void
+thread_event_update(tsd_t *tsd) {
+	uint64_t wait = thread_allocated_next_event_compute(tsd);
+	thread_event_adjust_thresholds_helper(tsd, wait);
+
+	uint64_t last_event = thread_allocated_last_event_get(tsd);
```

Comment:
Just realized, should we always push last event to current (in other words, do an event trigger) when updating? Otherwise, if thread allocated != last event, we can overcount?

Reply:
We currently only do `thread_event_trigger()` if `thread_allocated` has advanced too much i.e. surpassing `thread_allocated_next_event`. If it's still below `next_event`, I think it's fine to leave things as is, i.e. no event has triggered or should have triggered.

What do you mean by overcount? Based on my understanding, `thread_allocated != thread_allocated_last_event` may be true only in the following two (rare) scenarios:

- Reentrancy: interrupting calls advanced `thread_allocated`.
- `thread_event_update()` is called without being proceeded by `thread_event()`, e.g. in `prof_reset()` and other one-time circumstances.

I don't think either scenario involves any overcounting?

Reply:
By overcount, I mean when `allocated != last_event`, if we do event_update to refresh a single event, it may get inserted w/o touching last_event. In that case, next time we trigger an event, the accum_bytes for that refreshed event would be more than what actually happened since it got updated.

I agree those should be rare, as in the common case the event_update should be right after the event_trigger. Let's still find a reasonable fix for it though.

Reply:
> By overcount, I mean when `allocated != last_event`, if we do event_update to refresh a single event, it may get inserted w/o touching last_event. In that case, next time we trigger an event, the accum_bytes for that refreshed event would be more than what actually happened since it got updated.

I think by not touching `thread_allocated_last_event`, we're actually achieving the goal of making the computation right at the next event triggering time.

Say at the start of the `thread_event_update()` call, `thread_allocated_last_event` is 100, `thread_allocated` is 300, and the event wait time is 1000, and the caller of `thread_event_update()` is trying to set 800 as the new event wait time. If we don't touch `thread_allocated_last_event`, the event will be triggered when `thread_allocated` reaches 900, meaning that we're counting the 200 bytes that have been already allocated and we only need 600 bytes more, which sounds right. If we instead set `thread_allocated_last_event` to 300, then we're effectively discarding the 200 bytes that have already been allocated, and won't trigger the event until `thread_allocated` reaches 1100 - this is the same as keeping the wait time as 1000 so the caller effectively failed to update the wait time.

Reply:
Ah I see your point. You convinced me: counting all bytes since the event wait dropped to 0 makes sense.